### PR TITLE
Add autoencoder support to behavior cloning training

### PR DESCRIPTION
**Pull request by sivral**
## Description
This PR integrates a new vision-based autoencoder class into the behavior cloning training.
The autoencoder includes a decoder that reconstructs images from the encoder’s latent representation. When the autoencoder is selected in the BC training configuration, an additional reconstruction loss is computed alongside action and pose losses.

## Motivation and context
The motivation for this change is to evaluate the effectiveness of visual state representations learned via autoencoders in BC training, and to simply provide a new method for vision-based BC.

## How has this been tested?
Manually, on Geonosis PC.

## Checklist
- [x] All TODOs in the code have been resolved or linked to a proper issue.
- [x] Code has been (auto)formatted.
- [x] Documentation (e.g., README, CHANGELOG, Wiki) has been updated.
- [x] All automated checks have passed.

---
Clickup task: [869cbu87z](https://app.clickup.com/t/869cbu87z)

### CODE REVIEW

File: aegis_gym/rsl/behavior_cloning.py [hunk 1]
```diff
@@ -84,6 +84,10 @@ def __init__(
 
     def learn(self, num_learning_iterations: int) -> None:
         self._buffer.clear()
+        encoder_type = self._cfg["policy"]["type"]
+        use_autoencoder = encoder_type == "autoencoder"
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Right now this key name is confusing. 
```python
        encoder_type = self._cfg["policy"]["perception_type"]
        use_autoencoder = encoder_type == "autoencoder"
```

File: aegis_gym/rsl/behavior_cloning.py [hunk 2]
```diff
@@ -84,6 +84,10 @@ def __init__(
...
+        save_recons = True
+        save_recon_freq = 10
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Shouldn't these values be in the policy's config?

File: aegis_gym/rsl/behavior_cloning.py [hunk 3]
```diff
@@ -110,14 +118,25 @@ def learn(self, num_learning_iterations: int) -> None:
...
-                pose_loss = 0.0
+                pose_loss = th.tensor(0.0, device=self._device)
                 for pred_pose in pred_poses:
                     pose_loss += self._compute_pose_loss(
                         pred_pose, batch["object_poses"]
                     )
 
+                recon_loss = th.tensor(0.0, device=self._device)
+                if use_autoencoder:
+                    recons = self._policy.vision_encoder.reconstruct(batch["rgb_obs"])
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: I would recommend to access the vision encoder via getter - it will be more future proofed for modifications:

```python
                    enc =  self._policy.get_encoder()
                    recons = enc.reconstruct(batch["rgb_obs"])
```

File: aegis_gym/rsl/behavior_cloning.py [hunk 4]
```diff
@@ -129,8 +148,17 @@ def learn(self, num_learning_iterations: int) -> None:
...
                 num_batches += 1
 
+            if (
+                use_autoencoder
+                and save_recons
+                and (it + 1) % save_recon_freq == 0
+                and last_recons is not None
+            ):
+                self._save_reconstructions(last_batch, last_recons, it)
+
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: move this logic (i.e. `if`+call) and the save call to a new auxiliary method

File: aegis_gym/rsl/behavior_cloning.py [hunk 5]
```diff
@@ -237,20 +267,35 @@ def _collect_with_rl_teacher(self) -> None:
                 )
                 self._cur_reward_sum[new_ids] = 0
 
+    def _save_reconstructions(self, batch, recons, it):
+        import torchvision.utils as vutils
+
+        os.makedirs("reconstructions", exist_ok=True)
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Please switch to the Pathlib:

```python
        Path("reconstructions").mkdir(exist_ok=True)
```

File: aegis_gym/rsl/behavior_cloning.py [hunk 6]
```diff
@@ -237,20 +267,35 @@ def _collect_with_rl_teacher(self) -> None:
                 )
                 self._cur_reward_sum[new_ids] = 0
 
+    def _save_reconstructions(self, batch, recons, it):
+        import torchvision.utils as vutils
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: There is no reason to call multiple imports during the runtime. Please move it to the imports section.

File: aegis_gym/rsl/behavior_cloning.py [hunk 7]
```diff
@@ -237,20 +267,35 @@ def _collect_with_rl_teacher(self) -> None:
...
+
+        os.makedirs("reconstructions", exist_ok=True)
+
+        b = 0
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: What is be? `becon`? `batch`? Unfold it please

File: aegis_gym/rsl/behavior_cloning.py [hunk 8]
```diff
@@ -237,20 +267,35 @@ def _collect_with_rl_teacher(self) -> None:
...
+        for c in range(self._policy.num_cameras):
+            orig = batch["rgb_obs"][b, c * 3 : (c + 1) * 3].detach().cpu()
+            recon = recons[c][b].detach().cpu()
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Hm, the `c` sounds good for both `cameras` and `channels`. Please update the lines 130-133 to match this convention

File: aegis_gym/rsl/behavior_cloning.py [hunk 9]
```diff
@@ -237,20 +267,35 @@ def _collect_with_rl_teacher(self) -> None:
...
+
+            vutils.save_image(orig, f"reconstructions/orig_iter{it + 1:04d}_c{c}.png")
+            vutils.save_image(recon, f"reconstructions/recon_iter{it + 1:04d}_c{c}.png")
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py:307]: Please log them to the ClearML to the "Debug Samples" tab:
```python
from clearml import Task

task = Task.init(project_name="your_project", task_name="your_training_task")
logger = task.get_logger()

# ....

generated_img = ...  # your generated image from DL model or path
logger.report_image(
        title="Generated Samples",      # groups images in UI
        series="Reconstructions",        # sub-group/series name
        iteration=iteration,             # training step/epoch
        image=generated_img              # np.array, PIL.Image, or path
    )
```

File: aegis_gym/rsl/behavior_cloning.py [hunk 10]
```diff
@@ -472,11 +517,17 @@ def cnn_builder() -> nn.Sequential:
...
                 vision_cfg=vision_cfg,
             )
+        elif encoder_type == "autoencoder":
+            return AutoencoderCNNEncoder(
+                self.num_cameras,
+                cnn_builder=cnn_builder,
+                vision_cfg=vision_cfg,
+            )
         else:
             raise ValueError(f"Unknown vision encoder type: {encoder_type}")
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Change `elif` strutcture to `match`-`case`

File: aegis_gym/rsl/behavior_cloning.py [hunk 11]
```diff
@@ -491,9 +542,16 @@ def _build_cnn(config: dict) -> nn.Sequential:
...
 
-        if config.get("pooling") == "adaptive_avg":
-            pool_size = config["pool_size"]
-            layers.append(nn.AdaptiveAvgPool2d((pool_size, pool_size)))
+        if encoder_type != "autoencoder":
+            if config.get("pooling") == "adaptive_avg":
+                layers.append(
+                    nn.AdaptiveAvgPool2d((config["pool_size"], config["pool_size"]))
+                )
+            # TODO(issue#79): Remove hardcoded dimensions related to autoencoder and make them configurable
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Instead of nested ifs try to minimize indentation with logic negation:

```python
        if encoder_type == "autoencoder":
            return

        if config.get("pooling") == "adaptive_avg":
            layers.append(
                nn.AdaptiveAvgPool2d((config["pool_size"], config["pool_size"]))
            )
        # TODO(issue#79): Remove hardcoded dimensions related to autoencoder and make them configurable
        # ...
```

File: aegis_gym/rsl/behavior_cloning.py [hunk 12]
```diff
@@ -559,3 +617,66 @@ def forward(self, rgb_obs: th.Tensor) -> tuple[th.Tensor, ...]:
...
+
+        self.flatten_dim = self.feature_channels * self.feature_size * self.feature_size
+        self.latent_dim = 512
+
+        self.to_latent = nn.Linear(self.flatten_dim, self.latent_dim)
+        self.from_latent = nn.Linear(self.latent_dim, self.flatten_dim)
+
+        self.output_dim = self.latent_dim
+
+        self.decoder = nn.Sequential(
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: I know its hard coded, but lets start lying fundamentals for better design.

Prepare an new method, lets call it `build_decoder()`, and return this nn.Sequential network.

In next iteration we will dissect the cnn_builder() into something more modular.

File: aegis_gym/rsl/behavior_cloning.py [hunk 13]
```diff
@@ -559,3 +617,66 @@ def forward(self, rgb_obs: th.Tensor) -> tuple[th.Tensor, ...]:
...
+
+        for i in range(self.num_cameras):
+            cam_rgb = rgb_obs[:, i * 3 : (i + 1) * 3]
+
+            fmap = self.encoder(cam_rgb)
+            flat = fmap.flatten(start_dim=1)
+
+            latent = self.to_latent(flat)
+
+            features.append(latent)
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: There is really no reason to not pre-allocate list in this scenario:

```suggestion
        features = [None] * self.num_cameras

        for i in range(self.num_cameras):
            cam_rgb = rgb_obs[:, i * 3 : (i + 1) * 3]
            flat_fmap = self.encoder(cam_rgb).flatten(start_dim=1)
            features[i] = self.to_latent(flat_fmap)
```

File: aegis_gym/rsl/behavior_cloning.py [hunk 14]
```diff
@@ -559,3 +617,66 @@ def forward(self, rgb_obs: th.Tensor) -> tuple[th.Tensor, ...]:
...
+            flat = fmap.flatten(start_dim=1)
+
+            latent = self.to_latent(flat)
+            flat_recon = self.from_latent(latent)
+
+            fmap_recon = flat_recon.view(-1, 32, 16, 16)
+
+            recon = self.decoder(fmap_recon)
+
+            recons.append(recon)
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Similar here:


```suggestion
        recons = [None] * self.num_cameras

        for i in range(self.num_cameras):
            cam_rgb = rgb_obs[:, i * 3 : (i + 1) * 3]

            flat_fmap = self.encoder(cam_rgb).fmap.flatten(start_dim=1)
            latent = self.to_latent(flat_fmap)
            
            flat_recon = self.from_latent(latent)
            fmap_recon = flat_recon.view(-1, 32, 16, 16)
            
            recons[i] = self.decoder(fmap_recon)
```

**Review by macmacal — CHANGES_REQUESTED** [review]
Please address all comments

**Review by macmacal — CHANGES_REQUESTED** [review]
Looking good. Please continue with further changes. Please update the CHANGELOG and the PR description.

File: aegis_gym/rsl/behavior_cloning.py [hunk 15]
```diff
@@ -237,20 +269,33 @@ def _collect_with_rl_teacher(self) -> None:
                 )
                 self._cur_reward_sum[new_ids] = 0
 
+    def _save_reconstructions(self, batch, recons, it):
+        Path("reconstructions").mkdir(exist_ok=True)
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py:298]: Are you sure you haven't forget to import it via `from pathlib import Path`?

**Review by macmacal — CHANGES_REQUESTED** [review]

File: aegis_gym/rsl/behavior_cloning.py (hunk 1 above)
Comment (macmacal) [reply aegis_gym/rsl/behavior_cloning.py]: This is still relevant, please address it

File: aegis_gym/rsl/behavior_cloning.py (hunk 4 above)
Comment (macmacal) [reply aegis_gym/rsl/behavior_cloning.py]: Still relevant, please address it

File: aegis_gym/rsl/behavior_cloning.py [hunk 16]
```diff
@@ -110,14 +119,26 @@ def learn(self, num_learning_iterations: int) -> None:
...
                     )
 
+                recon_loss = th.tensor(0.0, device=self._device)
+                if use_autoencoder:
+                    enc = self._policy.get_encoder()
+                    recons = enc.reconstruct(batch["rgb_obs"])
+                    for c in range(self._policy.num_cameras):
+                        target = batch["rgb_obs"][:, c * 3 : (c + 1) * 3]
+                        recon_loss += F.mse_loss(recons[c], target)
+                    recon_loss /= self._policy.num_cameras
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: An idea: since autoencoder is an inherit property of the constructed nn.Module, set the `use_autoencoder` as a member variable. Then, we can move the calculation of the reconstruction loss to an auxiliary method without any additional arguments.

```
    recon_loss = calculate_recon_loss()

# ....
def calculate_recon_loss(self) -> th.Tensor:
    recon_loss = th.tensor(0.0, device=self._device)
    if not self._use_autoencoder:
        return recon_loss
    
    enc = self._policy.get_encoder()    
    recons = enc.reconstruct(batch["rgb_obs"])
    for c in range(self._policy.num_cameras):
        target = batch["rgb_obs"][:, c * 3 : (c + 1) * 3]
        recon_loss += F.mse_loss(recons[c], target)
        recon_loss /= self._policy.num_cameras
    return recon_loss
```

File: aegis_gym/rsl/behavior_cloning.py [hunk 17]
```diff
@@ -140,6 +170,7 @@ def learn(self, num_learning_iterations: int) -> None:
             else:
                 avg_action_loss = total_action_loss / num_batches
                 avg_pose_loss = total_pose_loss / num_batches
+                avg_recon_loss = total_recon_loss / num_batches
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Here it is also possible to simplify this logic by removing the else statemnt - it dosen't matter what happens after raising the `ValueError`:

```
if num_batches == 0:
    raise ValueError("No batches collected")

avg_action_loss = total_action_loss / num_batches
avg_pose_loss = total_pose_loss / num_batches
avg_recon_loss = total_recon_loss / num_batches
```

File: aegis_gym/rsl/behavior_cloning.py (hunk 15 above)
Comment (macmacal) [reply aegis_gym/rsl/behavior_cloning.py:298]: Checked - there is an import

File: aegis_gym/rsl/behavior_cloning.py (hunk 7 above)
Comment (macmacal) [reply aegis_gym/rsl/behavior_cloning.py]: It is still relevant

File: aegis_gym/rsl/behavior_cloning.py (hunk 9 above)
Comment (macmacal) [reply aegis_gym/rsl/behavior_cloning.py:307]: What is the status of this?

File: aegis_gym/rsl/behavior_cloning.py (hunk 11 above)
Comment (macmacal) [reply aegis_gym/rsl/behavior_cloning.py]: It is still relevant

**Review by macmacal — CHANGES_REQUESTED** [review]
There are still ongoing things which should be resolved.

File: aegis_gym/rsl/behavior_cloning.py (hunk 16 above)
Comment (macmacal) [reply aegis_gym/rsl/behavior_cloning.py]: This is still relevant

File: aegis_gym/rsl/behavior_cloning.py [hunk 18]
```diff
@@ -237,20 +263,44 @@ def _collect_with_rl_teacher(self) -> None:
                 )
                 self._cur_reward_sum[new_ids] = 0
 
+    def _save_reconstructions(self, batch, recons, it):
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Add typings

File: aegis_gym/rsl/behavior_cloning.py [hunk 19]
```diff
@@ -237,20 +263,44 @@ def _collect_with_rl_teacher(self) -> None:
...
+        if (
+            self._enable_recon
+            and self._save_recons
+            and (it + 1) % self._save_recon_freq == 0
+            and recons is not None
+        ):
```
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Reverse logic to minimize indentation:
```python3
        recon_enabled = ( self._enable_recon
            and self._save_recons
            and (it + 1) % self._save_recon_freq == 0
            and recons is not None)
        if not recon_enabled:
            return
        
```

File: aegis_gym/rsl/behavior_cloning.py (hunk 18 above)
Comment (macmacal) [inline aegis_gym/rsl/behavior_cloning.py]: Add typings

**Review by macmacal — CHANGES_REQUESTED** [review]

File: aegis_gym/rsl/behavior_cloning.py (hunk 9 above)
Comment (macmacal) [reply aegis_gym/rsl/behavior_cloning.py:307]: Created issue #81 for it

**Review by macmacal — APPROVED** [review]
LGTM 🆗 , go for it

> [!IMPORTANT]
> PLEASE DO THE SQUASH MERGE
